kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/utils/TestUtils.scala

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.utils
+
+import java.util.{Optional, TimeZone}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+import io.delta.kernel.{Scan, Snapshot, Table}
+import io.delta.kernel.client.TableClient
+import io.delta.kernel.data.Row
+import io.delta.kernel.defaults.client.DefaultTableClient
+import io.delta.kernel.types._
+import io.delta.kernel.utils.CloseableIterator
+import org.apache.hadoop.conf.Configuration
+import org.scalatest.exceptions.TestFailedException
+import org.scalatest.Assertions
+
+trait TestUtils extends Assertions {
+
+  lazy val defaultTableClient = DefaultTableClient.create(new Configuration())
+
+  implicit class CloseableIteratorOps[T](private val iter: CloseableIterator[T]) {
+
+    def forEach(f: T => Unit): Unit = {
+      try {
+        while (iter.hasNext) {
+          f(iter.next())
+        }
+      } finally {
+        iter.close()
+      }
+    }
+
+    def toSeq: Seq[T] = {
+      try {
+        val result = new ArrayBuffer[T]
+        while (iter.hasNext) {
+          result.append(iter.next())
+        }
+        result
+      } finally {
+        iter.close()
+      }
+    }
+  }
+
+  implicit class StructTypeOps(schema: StructType) {
+
+    def withoutField(name: String): StructType = {
+      val newFields = schema.fields().asScala
+        .filter(_.getName != name).asJava
+      new StructType(newFields)
+    }
+  }
+
+  def readSnapshot(
+    snapshot: Snapshot,
+    readSchema: StructType = null,
+    tableClient: TableClient = defaultTableClient): Seq[Row] = {
+
+    val result = ArrayBuffer[Row]()
+
+    var scanBuilder = snapshot.getScanBuilder(tableClient)
+
+    if (readSchema != null) {
+      scanBuilder = scanBuilder.withReadSchema(tableClient, readSchema)
+    }
+
+    val scan = scanBuilder.build()
+
+    val scanState = scan.getScanState(tableClient);
+    val fileIter = scan.getScanFiles(tableClient)
+    // TODO serialize scan state and scan rows
+
+    fileIter.forEach { fileColumnarBatch =>
+      // TODO deserialize scan state and scan rows
+      val dataBatches = Scan.readData(
```

Comment:
close `dataBatches`

Reply:
It's closed as part of the `.forEach` here https://github.com/delta-io/delta/pull/2034/files/7905c4d8e2d3c66d1a78e37c05cb6b02211cf9aa#diff-4b532e689ace6208898c69a5aa728bb7e76f97d08a39cc1632946c59f73f56a3R39